crypto/signer.go

```diff
@@ -0,0 +1,83 @@
+package neofscrypto
+
+import (
+	"fmt"
+)
+
+// Scheme represents digital signature algorithm with fixed cryptographic hash function.
+//
+// Non-positive values are reserved and depend on context (e.g. unsupported scheme).
+type Scheme uint32
+
+//nolint:revive
+const (
+	_ Scheme = iota
+
+	ECDSA_SHA512               // ECDSA with SHA-512 hashing (FIPS 186-3)
+	ECDSA_DETERMINISTIC_SHA256 // Deterministic ECDSA with SHA-256 hashing (RFC 6979)
+)
+
+// maps Scheme to blank PublicKey constructor.
+var publicKeys = make(map[Scheme]func() PublicKey)
+
+// RegisterScheme registers a function that returns a new blank PublicKey
+// instance for the given Scheme. This is intended to be called from the init
+// function in packages that implement signature schemes.
+//
+// RegisterScheme panics if function for the given Scheme is already registered.
+func RegisterScheme(scheme Scheme, f func() PublicKey) {
+	_, ok := publicKeys[scheme]
+	if ok {
+		panic(fmt.Sprintf("scheme %v is already registered", scheme))
+	}
+
+	publicKeys[scheme] = f
+}
+
+// Signer is an interface of entities that can be used for signing operations
+// in NeoFS. Unites secret and public parts. For example, an ECDSA private key
+// or external auth service.
+//
+// See also PublicKey.
+type Signer interface {
+	// Scheme returns corresponding signature scheme.
+	Scheme() Scheme
+
+	// Sign signs digest of the given data. Implementations encapsulate data
+	// hashing that depends on Scheme. For example, if scheme uses SHA-256, then
+	// Sign signs SHA-256 hash of the data.
+	Sign(data []byte) ([]byte, error)
+
+	// MaxPublicKeyEncodedSize returns maximum size required for binary-encoded
+	// public key.
+	//
+	// MaxPublicKeyEncodedSize MUST NOT return value greater than any return of
+	// EncodePublicKey.
+	MaxPublicKeyEncodedSize() int
+
+	// EncodePublicKey encodes public key into buf. Returns number of bytes
+	// written.
+	//
+	// EncodePublicKey MUST panic if buffer size is insufficient and less than
+	// MaxPublicKeyEncodedSize (*). EncodePublicKey MUST return negative value
+	// on any failure except (*).
+	//
+	// EncodePublicKey is expected to be compatible with PublicKey.Decode for
+	// similar signature schemes.
+	EncodePublicKey(buf []byte) int
```

Comment:
I think we should follow `ecdsa` and `crypto` interface here, `Signer.PublicKey` would be convenient to sign something and use a public key for _something_ without `Signer.EncodePublicKey() + PublicKey.Decode()` steps. And these 2 methods can be moved to the `PublicKey` itself, removing the need to specify `PublicKey` in their names.

Reply:
Done.